addons/mail/controllers/thread.py

```diff
@@ -23,14 +23,37 @@ def mail_thread_data(self, thread_model, thread_id, request_list, **kwargs):
             ).get_result()
         return Store(thread, as_thread=True, request_list=request_list).get_result()
 
-    @http.route("/mail/thread/messages", methods=["POST"], type="json", auth="user")
-    def mail_thread_messages(self, thread_model, thread_id, search_term=None, before=None, after=None, around=None, limit=30):
-        domain = [
-            ("res_id", "=", int(thread_id)),
+    def _get_fetch_domain(self, thread_model, thread_id, **kwargs):
+        return [
+            ("res_id", "=", thread_id),
             ("model", "=", thread_model),
             ("message_type", "!=", "user_notification"),
         ]
-        res = request.env["mail.message"]._message_fetch(domain, search_term=search_term, before=before, after=after, around=around, limit=limit)
+
+    @http.route("/mail/thread/messages", methods=["POST"], type="json", auth="public")
+    def mail_thread_messages(
+        self,
+        thread_model,
+        thread_id,
+        search_term=None,
+        before=None,
+        after=None,
+        around=None,
+        limit=30,
+        **kwargs
+    ):
+        thread = request.env[thread_model]._get_thread_with_access(thread_id, **kwargs)
+        if not thread:
+            raise Forbidden()
+        domain = self._get_fetch_domain(thread_model, int(thread_id), **kwargs)
+        # sudo: mail.thread - reading messages on accessible thread is allowed
+        res = (
+            thread.sudo()
+            .env["mail.message"]
+            ._message_fetch(
+                domain, search_term=search_term, before=before, after=after, around=around, limit=limit
```

Comment:
When I see this, I'm thinking `_message_fetch` should be a method on thread rather than on message.
Then the "thread part" of the domain can be automatically built-in.

Reply:
There are some other uses of `_message_fetch` like searching the inbox, history or starred messages that don't use the thread data in the domain. So maybe we want to keep it the way it is?